File: src/api/instance.ts
import { type Request, defaultFetcher, type Fetcher } from "@literate.ink/utilities";
import { cleanURL } from "./helpers/clean-url";
import { decodeInstance } from "~/decoders/instance";
import type { Instance } from "~/models";

export const instance = async (url: string, fetcher: Fetcher = defaultFetcher): Promise<Instance> => {
  url = cleanURL(url);
  url += "/infoMobileApp.json?id=0D264427-EEFC-4810-A9E9-346942A862A4";

  const request: Request = { url: new URL(url) };
  const response = await fetcher(request);
  const instance = JSON.parse(response.content);
  return decodeInstance(instance);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/instance.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.helpers.clearURL
import ink.literate.pawnote.decoders.decodeInstance
import ink.literate.pawnote.models.Instance
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject

suspend fun instance(url: String, httpClient: HttpClient = HttpClient()): Instance {
  var cleanedURL = clearURL(url)
  cleanedURL += "/infoMobileApp.json?id=0D264427-EEFC-4810-A9E9-346942A862A4"

  val response = httpClient.get(cleanedURL)
  return decodeInstance(Json.parseToJsonElement(response.bodyAsText()).jsonObject)
}
